src/fpm_lock.f90

```
@@ -0,0 +1,330 @@
+! that the lock-file wasn't left behind on accident. However this adds quite a
+! bit of complexity to the code, and it is very difficult / maybe even
+! impossible to do without race-conditions: There is no atomic way to tell the
+! OS to:
+!    > Open this file for reading with a shared lock in case it exists and
+!    > if it doesn't exists then create the file with for writing with an
+!    > exclusive lock.
+! Even if we we're able to do it there is a conceptual flaw: Distributed file
+! systems. If the package lives on another machine that you're accessing through
+! something like NFS then you'll end up writing your PID to a file living on
+! another machine that's not running the process, and so processes on that
+! machine can't know whether the lock is valid or not.
+!
+! Now we might turn to actual OS file-locking primitives such as `fcntl` on UNIX
+! and `LockFile` on Windows. This is again a step-up in complexity, and I don't
+! know about `LockFile` but `fcntl` is fraught with problems:
+! https://chris.improbable.org/2010/12/16/everything-you-never-wanted-to-know-about-file-locking/
+!
+! My conclusion is that anything more advanced than the current implementation
+! might just not be worth it, but I'm happy to be proven wrong! :-)
+
+module fpm_lock
+
+use :: fpm_error, only : error_t, fatal_error
+use :: fpm_os, only : get_current_directory
+use :: fpm_filesystem, only : join_path
+use, intrinsic :: iso_fortran_env, only : stderr => error_unit
+use iso_c_binding, only : c_int, c_char, c_null_char, c_ptr, c_funptr, &
+                          c_funloc, c_f_pointer
+use fpm_strings, only: f_string
+
+
+implicit none
+private
+public :: fpm_lock_acquire, fpm_lock_acquire_noblock, fpm_lock_release
+
+logical :: has_lock = .false.
+logical :: has_atexit_handler = .false.
+
+interface
+    ! This function is defined in `fpm_lock.c`.
+    subroutine c_create(path, iostat, iomsg, exists) bind(c, name='c_create')
```

Regarding `c` sources, I would advise to not use them whenever possible. If C is only used to create or delete a file, I think we should do it from Fortran. (see https://github.com/zoziha/MINGW-packages/pull/1 for a recent issue)

The reason is that `fpm` is distributed as a single-file `.F90` source, that is used to "bootstrap" fpm (fpm builds fpm). In such version, the C functions are not available (in your case, they should be safeguarded with `#ifdef FPM_BOOTSTRAP`). That is OK for simple builds, but could lead to unintended behavior during bootstrapping. 

So I would advise that we replace these C functions with Fortran equivalent. Please look at the `fpm_filesystem` and `fpm_os` modules, and if there is missing functionality, you can look into the Fortran `stdlib` implementation, or other open-source packages i.e. [M_io](https://github.com/urbanjost/M_io) (@urbanjost might be willing to help)

---

Thank you for pointing this out, I wasn't aware of this complication.

Have I understood you correctly that binding to C isn't problematic in this regard, rather, it's writing custom C, so the problematic code is `c_create` and `c_remove`, but not `free` and `atexit`?

I think `c_remove` can be replaced by calls to frotran's `open` and `close` (it will introduce a race condition that doesn't exist in `c_remove`, but that race condition is nothing more but slightly annoying to the programmer, it won't cause bugs)

`c_create` is more difficult though, the reason I feel that I "need" `c_create` is to avoid a problematic race-condition: We want to create `.fpm-package-lock` _only_ if it doesn't exist. A naïve approach that can be done in Fortran is something like this:
```
if (not_exists('.fpm-package-lock')) then
    open(file='.fpm-pakage-lock', ...) // has lock
else
   ... // does not have lock
end if
```
However, another Fortran process might create the lock-file after the `not_exists` check but before the `create`, and so now two processes think that they have a lock at the same time.

Another approach is to call `open` without first having a `not_exists` check, and then try to discern whether or not the file was created or simply opened, but I don't think this is possible (See this [note](https://github.com/emmabastas/fpm/blob/f1f14e49319554209f2a33ac5595ebe419c14fe3/src/fpm_lock.f90#L192))

# Ways forward

- `c_remove`. Remove usage of this function, pretty straightforward
- `c_create`. Here I can see some alternatives
    1. Implement this in Fortran and accept the race condition.
    2. Make it so that bootstrap:ed builds don't have the lock-file feature at all.
    3. Find some clever way to implement this in Fortran without a race condition.
    4. Call the C function `open` directly from Fortran instead of via `c_create`.
          I think this is doable. Maybe we'll lose the ability to get the error code/message if something goes wrong on the C side, but that's not the end of the world. It can also be finicky to find the right Fortran integers corresponding to  `O_RDONLY`, `O_CREAT`, etc.

---

```Fortran
open(file='.fpm-package-lock', ..., action='write', status='NEW', iostat=ios, iomsg=msg)
if (ios/=0) stop 'file exists or is not readable: '//trim(msg)
```

Should be enough for a single-call evaluation imho. So yes, I believe all the C calls in the current proposal should be replaceable with Fortran

---

Yes you are totally right that will work.. Idk why I got it in my that that that wasn't possible, anyways I will see to it.

---

Yes it would definitely be nice to differentiate between the two failures. Both the error code (`iostat`) and message (`iomsg`) will be compiler dependent though. For example [XLF](https://www.cenapad.unicamp.br/parque/manuais/Xlf/lr88.HTM) has `iostat==107`. Mapping all possible compiler/error code combinations does not seem feasible. Perhaps one option would be to call `inquire(filename,exist=is_old)`, to check if the file existed, only after we know that file creation failed?

---

- Agree that mapping compiler/error codes is not feasible
- If we do `inquire` I think we'll en up having a race condition; Something may happen to the lock-file after the `open(..)` that tries to create the lock-file but before the `inquire(..)` that checks if the lock-file exists already. Not sure how problematic this is, at the very least we might end up still giving the user the wrong idea of what the issue is.

I wonder I a pragmatic solution would be to just report the `iomsg` in the warning message we give to the user when a package-lock fails, and it will be clear o them if something is wrong or not. Right now we warn:

> Warning: file .fpm-package-lock exists.
>
> Another process seams to be working on this package
> already and this process will wait for .fpm-package-lock
> to be removed before proceeding. If you think that a
> previous process crashed/terminated without removing
> .fpm-package-lock then you can try removing it manually.
>
> If the problem persists then please file a bug report
> at https://github.com/fortran-lang/fpm/issues

we could change it to

> Warning: An error occurred when trying to create .fpm-package-lock
> Message: `iomsg`
>
> This likely means that another process is working on this package
> already, in which case .fpm-package-lock already exists.
> This process will wait for .fpm-package-lock to be removed
> before proceeding. If you think that a previous process
>  crashed/terminated without removing .fpm-package-lock
>  then you can try removing it manually.
>
> If the problem persists, or if the error message indicates that
> some unexpected error occurred, then please file a bug report
> at https://github.com/fortran-lang/fpm/issues

---

I agree: for now, let's just ensure that the operation succeeded. If not, let's raise an error containing the OS message in `iomsg`. We can think of having a more nuanced error message later.